torch/distributions/dirichlet.py

```diff
@@ -80,3 +80,12 @@ def entropy(self):
         return (torch.lgamma(self.concentration).sum(-1) - torch.lgamma(a0) -
                 (k - a0) * torch.digamma(a0) -
                 ((self.concentration - 1.0) * torch.digamma(self.concentration)).sum(-1))
+
+    @lazy_property
+    def natural_params(self):
+        V1 = Variable(self.concentration.data, requires_grad=True)
+        return (V1, )
```

Comment:
Do we plan on having those methods be part of the public interface? They're doing things that aren't fully legal in autograd (rewrapping data in other Variables like this). The semantics are also unclear for me (why does the result always require grad)?

Reply:
I suppose these could be private, since I don't know of an application outside of `.entropy()` and `.kl_divergence()`; however there may be future applications.

The `.natural_params` need to set `requires_grad=True` internally because `.entropy()` and `.kl_divergence` both use `grad()` under the hood, even if the original variables do not require grad. Would you recommend wrapping in `Variable(-, requires_grad=True)` only in the case that the original variables do not require grad? Or is there a better practice we can follow?

Reply:
A better practice would be to return `self.concentration` and let the caller decide if they want to take grads or not

Reply:
Regardless of whether the user wants to take grads, `.entropy()` and `.kl_divergence()` need to use `grad()` under the hood; that is why we are wrapping this with `requires_grad=True`. That is, even if the user does not want to compute gradients of `Dirichlet(concentration).entropy()`, the Bregman divergence implementation uses `grad()` under the hood and must use a differentiable form of `concentration`, i.e. even if the result of `.entropy()` is not differentiable wrt the user's input. See implementation of `ExponentialFamily.entropy()` for clarification.

Reply:
I understand. I'm just saying that setting `requires_grad=True` should be a responsibility of the caller. If a particular implementation of `entropy()` needs it, then it needs to set the flag.

Reply:
Ok, if I understand correctly then we should simply fail in the following case?
```py
dist = Dirichlet(Variable(torch.ones(2), requires_grad=False))
Dirichlet.entropy(dist)          # Succeeds.
ExponentialFamily.entropy(dist)  # Should this succeed or fail?
```
@vishwakftw I think in this case we can simply add `requires_grad=True` to all the test examples? And maybe add a check in `.entropy()` and `.kl_divergence()`:
```py
nparams = self.natural_params
if not all(p.requires_grad for p in nparams):
    raise NotImplementedError(".entropy() requires all parameters to requires_grad")
```

Reply:
No, I meant this:
```python3
class ExponentialFamily(Distribution):
  def entropy(self):
    natural_params = [Variable(p, requires_grad=True) for p in self.natural_params]
    ...
```

Reply:
If this is frequent (setting `require_grad = True` on existing variables to enable `grad()`, how about adding a convenience method `attach()`? (that was also my suggestion in https://github.com/pytorch/pytorch/issues/1734)